### PR TITLE
td: more netsplit handling

### PR DESCRIPTION
**Pull request by danielnorberg**

### CODE REVIEW

**Comment by danielnorberg** [comment]
@komamitsu PTAL

File: digdag-standards/src/main/java/io/digdag/standards/operator/td/TdForEachOperatorFactory.java [hunk 1]
```diff
@@ -130,20 +142,32 @@ protected String startJob(TaskExecutionContext ctx, TDOperator op, String domain
...
+                return job.getResult(ite -> {
+                    List<Config> rows = new ArrayList<>();
+                    while (ite.hasNext()) {
+                        rows.add(row(columnNames, ite.next().asArrayValue()));
+                        if (rows.size() > Limits.maxWorkflowTasks()) {
+                            TaskLimitExceededException cause = new TaskLimitExceededException("Too many tasks. Limit: " + Limits.maxWorkflowTasks());
+                            throw new TaskExecutionException(cause, TaskExecutionException.buildExceptionErrorConfig(cause));
+                        }
                     }
+                    return rows;
```
Comment (komamitsu) [inline digdag-standards/src/main/java/io/digdag/standards/operator/td/TdForEachOperatorFactory.java]: Why don't you clear `retry` state (if exists) for later retrying?

File: digdag-standards/src/main/java/io/digdag/standards/operator/td/TdWaitOperatorFactory.java [hunk 2]
```diff
@@ -134,9 +145,24 @@ private String startJob(TDOperator op, String domainKey)
 
         private boolean fetchJobResult(TDJobOperator job)
         {
-            // TODO: handle netsplits
+            Config resultState = state.getNestedOrSetEmpty(RESULT);
+
+            Optional<ArrayValue> firstRow;
+            try {
+                firstRow = job.getResult(ite -> ite.hasNext() ? Optional.of(ite.next()) : Optional.absent());
+            }
```
Comment (komamitsu) [inline digdag-standards/src/main/java/io/digdag/standards/operator/td/TdWaitOperatorFactory.java]: We don't need to clear `retry` state here?

File: digdag-standards/src/main/java/io/digdag/standards/operator/td/TdWaitOperatorFactory.java [hunk 3]
```diff
@@ -134,9 +145,24 @@ private String startJob(TDOperator op, String domainKey)
...
+                }
+                int retry = resultState.get(RETRY, int.class, 0);
+                resultState.set(RETRY, retry + 1);
+                int interval = (int) Math.min(INITIAL_RESULT_FETCH_RETRY_INTERVAL * Math.pow(2, retry), MAX_RESULT_FETCH_RETRY_INTERVAL);
+                logger.warn("Failed to download result of job '{}', retrying in {} seconds", job.getJobId(), interval, e);
+                throw TaskExecutionException.ofNextPolling(interval, ConfigElement.copyOf(state));
+            }
 
-            Optional<ArrayValue> firstRow = job.getResult(ite -> ite.hasNext() ? Optional.of(ite.next()) : Optional.absent());
+            state.remove(RESULT);
```
Comment (komamitsu) [inline digdag-standards/src/main/java/io/digdag/standards/operator/td/TdWaitOperatorFactory.java]: [trivial] I'm not sure which is better, `resultState.remove(RETRY)` or `state.remove(RESULT)`...

**Comment by komamitsu** [comment]
I added some comments, but it overall looks good 👍

File: digdag-standards/src/main/java/io/digdag/standards/operator/td/TdForEachOperatorFactory.java [hunk 4]
```diff
@@ -130,20 +142,32 @@ protected String startJob(TaskExecutionContext ctx, TDOperator op, String domain
...
+                });
+            }
+            catch (UncheckedIOException | TDClientException e) {
+                if (isDeterministicClientException(e)) {
+                    throw new TaskExecutionException(e, TaskExecutionException.buildExceptionErrorConfig(e));
                 }
-                return rows;
-            });
+                int retry = resultState.get(RETRY, int.class, 0);
+                resultState.set(RETRY, retry + 1);
```
Comment (danielnorberg) [inline digdag-standards/src/main/java/io/digdag/standards/operator/td/TdForEachOperatorFactory.java]: Yeah, ideally we'd have some structured primitives that we could use to abstract away the retry state-machine.

File: digdag-standards/src/main/java/io/digdag/standards/operator/td/TdOperatorFactory.java [hunk 5]
```diff
@@ -294,27 +306,48 @@ private static String escapePrestoTableName(TableParam table)
...
-        if (downloadFile.isPresent()) {
+        try {
             j.getResult(ite -> {
                 try (BufferedWriter out = workspace.newBufferedWriter(downloadFile.get(), UTF_8)) {
                     addCsvHeader(out, j.getResultColumnNames());
-
                     while (ite.hasNext()) {
                         addCsvRow(out, ite.next().asArrayValue());
                     }
+                    downloadState.set(DONE, true);
```
Comment (danielnorberg) [inline digdag-standards/src/main/java/io/digdag/standards/operator/td/TdOperatorFactory.java]: That's not necessary but ok.

File: digdag-standards/src/main/java/io/digdag/standards/operator/td/TdWaitOperatorFactory.java (hunk 2 above)
Comment (danielnorberg) [inline digdag-standards/src/main/java/io/digdag/standards/operator/td/TdWaitOperatorFactory.java]: It's removed on line 165.

File: digdag-standards/src/main/java/io/digdag/standards/operator/td/TdWaitOperatorFactory.java (hunk 3 above)
Comment (danielnorberg) [inline digdag-standards/src/main/java/io/digdag/standards/operator/td/TdWaitOperatorFactory.java]: Fetching the job result might happen several times, not only once.

**Pull request merged**